src/openvslam/data/map_database.cc

```diff
@@ -71,6 +71,69 @@ std::vector<keyframe*> map_database::get_all_keyframes() const {
     return keyframes;
 }
 
+std::vector<keyframe*> map_database::get_close_keyframes(const Mat44_t& pose,
```

Comment:
Why not try to find all keyframes within a radius of the pose requested instead? That would make it so that you don't need to sort (which saves compute time) and also an arbitrary set parameter for the number of keyframes to take. Deciding on a window size of keyframes to try to use is a much more reasonable parameterization to think about. 

How are the keyframes decided in the normal re-localization mode? Are those based on distance or only on matching quality?

Reply:
> Why not try to find all keyframes within a radius of the pose requested instead?

When selecting closest keyframes to a given position, the orientation of camera in the each keyframe is no less important than its position in relation to the chosen pose. For example, I think the key-frame distant to `1m` from the selected pose might be more useful in terms of relocalization-by, than `10cm` one but looking in opposite direction. Therefore, I thinking that selecting the N-closest keyframes to relocalize by seems to be more stable.

> How are the keyframes decided in the normal re-localization mode? Are those based on distance or only on matching quality?

The normal relocalization is making by gathering relocalization keyframes candidates by BoW database (something like matching by features similarity `>=80%` as I undestand). In this relocalization the distance and mutual position of frame & candidate keyframe are clearly not taken into account explicitly.

> Whether here or in the ROS wrapper using it, it may be good to fallback to the normal global-relocalization method if this local initial pose setting fails to produce a usable track.

Good idea. Thanks for noting this.

Reply:
The logic of `tracking_module::track_current_frame()` was updated. Now, if the relocalization by pose failed, usual relocalization is being called: https://github.com/OpenVSLAM-Community/openvslam/pull/79/files#diff-9d7c6e92591d28cb8c87b3973729fddb3abd564e9feae8861851fd48dd31b2caR348

Reply:
How about the following implementation?

```c++
std::vector<keyframe*> map_database::get_close_keyframes(const Mat44_t& pose,
                                                         const double distance_threshold,
                                                         const double angle_threshold) const {
    std::lock_guard<std::mutex> lock(mtx_map_access_);

    const auto cos_angle_threshold = std::cos(angle_threshold);
    std::vector<std::pair<unsigned int, double>> cos_angles;
    std::vector<std::pair<unsigned int, double>> distances;
    std::vector<keyframe*> filtered_keyframes;

    // Calculate angles and distances between given pose and all keyframes
    Mat33_t M = pose.block<3, 3>(0, 0);
    Vec3_t Mt = pose.block<3, 1>(0, 3);
    for (const auto id_keyframe : keyframes_) {
        Mat33_t N = id_keyframe.second->get_cam_pose().block<3, 3>(0, 0);
        Vec3_t Nt = id_keyframe.second->get_cam_pose().block<3, 1>(0, 3);
        // Angle between two cameras related to given pose and selected keyframe
        const double cos_angle = ((M * N.transpose()).trace() - 1) / 2;
        // Distance between given pose and selected keyframe
        const double distance = (Nt - Mt).norm();
        if (distance < distance_threshold && cos_angle > cos_angle_threshold) {
            filtered_keyframes.push_back(keyframes_.at(item.second));
        }
    }

    return filtered_keyframes;
}
```

Reply:
I think here we are getting attached to each environment and how map was recorded. In initial implementation even if the angles and distances will be somehow above threshold, it still try to re-localize by pose. In the new implementation, the algorithm might be less stable and dependent from the initial map (how it was build, and where the keyframes are located in relation to requested pose). The new algorithm is much faster. But as, "initial_pose" requests are designed to came not frequently (usually a one or few times per localization session), I think we could we can sacrifice a little performance (algorithm speed).

However, from other hand it is apparently that starting from some threshold distance (say >=2 meters) and angles, the relocalization by pose won't be effective at all. And, as @SteveMacenski told, thresholds are more understandable value to developers who know nothing about OpenVSLAM internals. So, there might be a reason to change the searching algorithm to angle and distances thresholds. @ymd-stella, maybe you have something else to add? I still not fully sure it's worth changing something in search algorithm.

Reply:
> However, from other hand it is apparently that starting from some threshold distance (say >=2 meters) and angles, the relocalization by pose won't be effective at all.

So, whether we increase `limit` in the initial implementation or increase the threshold in a new implementation, they will behave in a similar way.
If the behavior is almost the same, the simpler (and faster) implementation is preferable.

Reply:
OK. Updated in the latest commit to using pose and cos_angle thresholds.